File: 338.counting-bits.py
#
# @lc app=leetcode id=338 lang=python3
#
# [338] Counting Bits
#

# @lc code=start
from typing import List


class Solution:
    def countBits(self, n: int) -> List[int]:
        #練習&位元運算， 只是比191題多加了for迴圈和一個串列而已
        res = []

        for i in range(n + 1):
            count = 0
            while (i != 0):
                i = i & (i - 1)
                count += 1
            res.append(count)
        return res
# @lc code=end

File: PracticeFolder/338.counting-bits.cs
/*
 * @lc app=leetcode id=338 lang=csharp
 *
 * [338] Counting Bits
 */

// @lc code=start
public class Solution {
    public int[] CountBits(int n)
    {
        //使用Bit Manipulation(位元運算) 
        int[] result = new int[n+1];
        for (int i = 0; i <= n; i++)
        {
            //要算i轉成2進制有幾個1，只要看當下i的上一個次方，再加上i除於2的餘數即可
            result[i] = result[i/2] + i%2;
        }
        return result;
    }
}
// @lc code=end

